### PR TITLE
Embed puzzle, sheet (if available), text, and voice channel links in discord messages

### PR DESCRIPTION
**Pull request by asdfryan**
Screen Shot 2021-01-13 at 5 39 36 PM

Both the Text and Voice links will auto-join the channels. 

The Sheet and Puzzle links will open a dialog to take you to an outside site.

One caveat is that because discord and sheets creation have both been celerified (and sheet creation is much slower than discord), the sheet will not be available on the first announcement.

I'll make a separate PR to announce (only on the puzzle-specific text channel) when the sheet has been created.

### CODE REVIEW

File: discord_lib/discord_chat_service.py [hunk 1]
```diff
@@ -21,12 +24,36 @@ def __init__(self, settings, client=None, max_channels_per_category=50):
         self._puzzle_announcements_id = settings.DISCORD_PUZZLE_ANNOUNCEMENTS_CHANNEL
         self._max_channels_per_category = max_channels_per_category
 
-    def send_message(self, channel_id, msg):
-        self._client.channels_messages_create(channel_id, content=msg)
-
-    def announce(self, msg):
+    @staticmethod
+    def __maybe_create_in_client_invite_link(url):
```
[the remark this answers is not in the document]
Comment (akirabaruah) [reply discord_lib/discord_chat_service.py]: Also, where do we use this function?

**Review by akirabaruah — CHANGES_REQUESTED** [review]
LGTM overall, with some small comments

File: discord_lib/discord_chat_service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (asdfryan) [reply discord_lib/discord_chat_service.py]: I deleted it :D

File: api/views.py [hunk 2]
```diff
@@ -101,7 +101,7 @@ def destroy(self, request, pk=None, **kwargs):
                 puzzle.status = Puzzle.SOLVING
                 if puzzle.chat_room:
                     transaction.on_commit(
-                        lambda: chat.tasks.handle_puzzle_unsolved.delay(puzzle_id)
+                        lambda: chat.tasks.handle_puzzle_unsolved.delay(puzzle.id)
```
Comment (rgossiaux) [inline api/views.py:104]: Oops, sorry, thanks for fixing this.

File: chat/models.py [hunk 3]
```diff
@@ -70,8 +70,8 @@ def create_channels(self):
         service = self.get_service()
         self.text_channel_id = service.create_text_channel(self.name)
         self.audio_channel_id = service.create_audio_channel(self.name)
-        self.text_channel_url = service.create_channel_url(self.text_channel_id)
-        self.audio_channel_url = service.create_channel_url(self.audio_channel_id)
+        self.text_channel_url = service.create_channel_url(self.text_channel_id, False)
+        self.audio_channel_url = service.create_channel_url(self.audio_channel_id, True)
```
Comment (rgossiaux) [inline chat/models.py:74]: nit: might be slightly clearer to use the `is_audio=` kwarg explicitly

File: puzzles/models.py [hunk 4]
```diff
@@ -175,6 +175,22 @@ def can_delete(self):
...
 
+    def create_field_url_map(self):
+        """
+        Creates dictionary mapping text -> url for all relevant urls if they exist.
+        """
+        field_url_map = {}
+        field_url_map["Puzzle"] = self.url
+        if self.chat_room:
+            if self.chat_room.text_channel_url:
+                field_url_map["Text"] = self.chat_room.text_channel_url
```
Comment (rgossiaux) [inline puzzles/models.py:186]: I feel like "Text" could be a little clearer but I don't actually have a good alternative. "Text chat"? "Text channel"? "Chat" (sounds confusing w/ voice chat but if there's a separate Voice button maybe it won't be confusing)? Not sure

File: chat/tasks.py [hunk 5]
```diff
@@ -27,8 +26,7 @@ def handle_puzzle_solved(puzzle_id, answer_text):
     try:
         puzzle.chat_room.archive_channels()
         msg = f"{puzzle.name} has been solved with {answer_text}!"
-        puzzle.chat_room.get_service().announce(msg)
-        puzzle.chat_room.send_message(msg)
+        puzzle.chat_room.send_and_announce_message_with_embedded_urls(msg, puzzle)
```
Comment (rgossiaux) [inline chat/tasks.py:29]: Do we actually need to include the embed here? Wondering if it's gonna feel a little spammy / unnecessary, what do you think?

File: chat/tasks.py [hunk 6]
```diff
@@ -38,6 +36,8 @@ def handle_puzzle_unsolved(puzzle_id):
     puzzle = Puzzle.objects.get(id=puzzle_id)
     try:
         puzzle.chat_room.unarchive_channels()
+        msg = f"{puzzle.name} is no longer solved!"
+        puzzle.chat_room.send_and_announce_message_with_embedded_urls(msg, puzzle)
```
Comment (rgossiaux) [inline chat/tasks.py:40]: Here too

File: puzzles/models.py (hunk 4 above)
Comment (asdfryan) [reply puzzles/models.py:186]: Renamed to "Text Channel" and "Voice Channel"

**Review by rgossiaux — APPROVED** [review]
Thanks! lgtm, think @akirabaruah  needs to approve

**Review by akirabaruah — APPROVED** [review]
This looks awesome. Thanks!

**Pull request merged**